### ISSUE TITLE
Issues encountered in scaling recent DCR on Piz Daint

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#1052]
This list is mostly for myself, to document issues encountered on recent scaling runs on Piz Daint. I will follow up separately for specific issues but hopefully this will help jog my memory so I know what to reproduce and how.

Note that these are symptoms, not every symptom is necessarily a separate runtime bug (or a runtime bug at all; there could be issues in different parts of the system).

 - [x] ~stencil/circuit/pennant: `nodcr_noidx` crashes with `bad_alloc` https://github.com/StanfordLegion/legion/issues/1049~
 - [x] ~stencil/circuit: `dcr_noidx` freezes at >= 512 nodes (note: not fixed by `-lg:window`)~
 - [x] ~stencil: `dcr_idx` gets wrong answer with >= 2 nodes unless you run with `-lg:no_fence_elision`~
 - [x] ~circuit: `dcr_noidx` gets (non-deterministic?) uninitialized read-only field error with >= 4 nodes (worked around by blanket fill on top-level regions)~
 - [ ] pennant: `dcr_noidx` is 10x slower than `dcr_idx` (probably an application or mapper issue)
 - [x] ~soleil: `dcr_noidx` freezes at 2 nodes when `-lg:sched -1` is used (maybe Legion should assert if this value is passed?)~
 - [x] ~soleil: `dcr_noidx` freezes at 16 nodes unless `-lg:window` (with a sufficiently large value) is used~
 - [ ] soleil: I've been seeing a number of distinct symptoms, depending on configuration and node count:
     - [ ] `hit_dom` in `base` configuration freezes at 1 (?) node (or at least 2)
     - [ ] `hit_dom` in `stag-200` with 1020/204 freezes at 16 nodes (at start of run)
     - [x] ~`hit_dom` in `stag-200` with 780/156 has a variety of symptoms:~
         - [x] ~non-deterministic crashes (related to distributed collectables)~
         - [x] ~non-deterministic freezes (sometimes at start, sometimes after iterations 199 or 399 (i.e., immediately before DOM))~
         - [x] ~deterministic shutdown hang starting at 16 nodes~
 - [ ] `REALM_NETWORKS=gasnetex` crash: https://github.com/StanfordLegion/legion/issues/1040
 - [ ] LLVM 6 performance regression: https://github.com/StanfordLegion/legion/issues/1041

**Comment by lightsighter** [comment]
There has been a fix applied to a hang in the future all reduce collective now, so all the tests that hung before need to be rerun.

**Comment by lightsighter** [comment]
> stencil: dcr_idx gets wrong answer with >= 2 nodes unless you run with -lg:no_fence_elision

Please provide a debug binary and command line to reproduce this one on sapling.

**Comment by elliottslaughter** [comment]
> There has been a fix applied to a hang in the future all reduce collective now, so all the tests that hung before need to be rerun.

This has fixed the stencil/circuit freeze at 512 nodes.

**Comment by elliottslaughter** [comment]
I have not been able to reproduce any of the Soleil symptoms at the 780/156 problem size. I'm running some more tests now to rule out the distributed collectable issue, but the shutdown hang was quite deterministic before, so I'm going to consider that gone.

I am still seeing freezes at the 1020/204 problem size. It may be that this genuinely is a memory issue. I'll need to investigate more to be sure.

**Comment by lightsighter** [comment]
> stencil: dcr_idx gets wrong answer with >= 2 nodes

This is fixed

> I am still seeing freezes at the 1020/204 problem size. It may be that this genuinely is a memory issue

Why do you think a memory issue would  produce a hang?

**Comment by elliottslaughter** [comment]
> > stencil: dcr_idx gets wrong answer with >= 2 nodes
> 
> This is fixed

Confirmed.

> > I am still seeing freezes at the 1020/204 problem size. It may be that this genuinely is a memory issue
> 
> Why do you think a memory issue would produce a hang?

No real reason except that the hang is problem size dependent and behavior can get finicky as you approach the node's memory limit, even if swap is disabled.

**Comment by lightsighter** [comment]
You also owe me a reproducer for the program that was giving you a read-only uninitialized error for sub-operations inside of a virtually mapped task in a control-replicated context.

**Comment by lightsighter** [comment]
> No real reason except that the hang is problem size dependent and behavior can get finicky as you approach the node's memory limit, even if swap is disabled.

Ok, yes, this does make sense if the runtime grinds to halt because malloc/free operations that it needs start going into swap space. Would be good to confirm that is occurring.

**Comment by lightsighter** [comment]
Also the `std::bad_alloc` is a duplicate of #1049 when assertions are off for release mode.

**Comment by elliottslaughter** [comment]
Binary and command line to replicate the uninitialized data issue:

```
/scratch2/eslaught/regent-index-launch-sc21/language/circuit.run1_debug_uninit
```

```
mpirun -n 2 -npernode 1 -bind-to none -x LD_LIBRARY_PATH=$PWD ./circuit.noidx -npp 500 -wpp 2000 -l 50 -p $(( 3 * 10 )) -pps 10 -prune 10 -hl:sched 1024 -ll:gpu 1 -ll:util 2 -ll:bgwork 2 -ll:csize 15000 -ll:fsize 15000 -ll:zsize 2048 -ll:rsize 512 -ll:gsize 0 -dm:replicate 1 -dm:same_address_space
```

Should be compatible with any recent `control_replication`.

**Comment by lightsighter** [comment]
@elliottslaughter Pull the most recent control replication. The uninitialized read-only error should be gone. You actually found the very first bug anyone has ever reported with replicated equivalence sets.

What hangs are still reproducible at this point?

**Comment by elliottslaughter** [comment]
There's the issue with `-lg:sched -1`, which I'm not sure is a valid option in the first place. We might solve that one by just disallowing values <= 0 on that command line flag (unless you've changed your mind about this being a valid value).

There are also the DOM freezes with the larger problem size, where I'm not sure if they're "real" or related to memory pressure.

**Comment by elliottslaughter** [comment]
Let me try again to see if I can get direct access to Piz Daint nodes. Their user guide claims that it's still possible, so maybe I just am not doing something right. If I can, I'll watch during a run and see if it's actually running close to the memory limit on the machine.

**Comment by lightsighter** [comment]
Ok, I think we should try `-lg:sched -1` and see if it reproduces after the other hang fix. Let me know if you find anything else that seems suspicious.

**Comment by elliottslaughter** [comment]
I don't seem to be seeing a freeze related to the presence of `-lg:sched -1`. I am still seeing a freeze if I fail to increase the `-lg:window` parameter. Was that something you wanted to address? I can add it if you want.

**Comment by lightsighter** [comment]
> I am still seeing a freeze if I fail to increase the -lg:window parameter. Was that something you wanted to address? I can add it if you want.

Yes, I still want to address it, but you don't need to look into it. I understand it. It should go away if you run with `-lg:no_physical_tracing`. Knowing how to reproduce it for when I push a fix would be good so we can test that it has been fixed.

**Comment by lightsighter** [comment]
There is now a fix in both master and control replication for the freeze that requires increasing the size of `-lg:window`. Please pull and try again to see if the freeze still manifests without the `-lg:window` option.

**Comment by elliottslaughter** [comment]
Confirmed fixed, thanks.

**Comment by lightsighter** [comment]
Do we think there are any more outstanding runtime issues that you need me to investigate?

**Comment by elliottslaughter** [comment]
I don't think so, and the important (Realm/Regent) bugs are in separate issues. I'll go ahead and close this.
